external/shellclient/v1/shellclient.go

```diff
@@ -0,0 +1,352 @@
+package v1
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/taskcluster-worker/plugins/interactive/shellconsts"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+)
+
+const (
+	// MaxOutstandingBytes is the maximum number of unsent bytes in the stdin buffer.
+	MaxOutstandingBytes = 8 * 1024 * 1024
+	pingInterval        = 15 * time.Second
+	writeTimeout        = 2 * pingInterval
+	readTimeout         = 3 * pingInterval
+)
+
+var dialer = websocket.Dialer{
+	HandshakeTimeout: shellconsts.ShellHandshakeTimeout,
+	ReadBufferSize:   shellconsts.ShellMaxMessageSize,
+	WriteBufferSize:  shellconsts.ShellMaxMessageSize,
+}
+
+// ShellClient is the implementation of
+// github.com/taskcluster/taskcluster-worker/engines.Shell for v1 shells.
+type ShellClient struct {
+	ws       *websocket.Conn
+	done     chan struct{}
+	mWrite   sync.Mutex
+	cPause   *sync.Cond
+	paused   bool
+	complete atomics.Once
+	success  bool
+	err      error
+
+	stdin        io.WriteCloser
+	stdinReader  io.ReadCloser
+	stdout       io.ReadCloser
+	stdoutWriter *DrainingPipeWriter
+	stderr       io.ReadCloser
+	stderrWriter *DrainingPipeWriter
+}
+
+// Dial will built a proper shell websocket URL, connect to the server and return an initialized shell client.
+func Dial(socketURL string, command []string, tty bool) (*ShellClient, error) {
+	u, err := url.Parse(socketURL)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid socketURL: %s, parsing error: %s",
+			socketURL, err)
+	}
+	q := u.Query()
+
+	// Ensure the URL has ws or wss as scheme
+	switch u.Scheme {
+	case "http":
+		u.Scheme = "ws"
+	case "https":
+		u.Scheme = "wss"
+	} // Set command arguments overwriting any existing querystring values
+	q.Del("command")
+	if len(command) > 0 {
+		for _, arg := range command {
+			q.Add("command", arg)
+		}
+	}
+
+	// Set tty true or false
+	if tty {
+		q.Set("tty", "true")
+	} else {
+		q.Set("tty", "false")
+	}
+
+	// Set querystring on url
+	u.RawQuery = q.Encode()
+
+	// Dial up to the constructed URL
+	ws, _, err := dialer.Dial(u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return a new ShellClient wrapping the websocket
+	return New(ws), nil
+}
+
+// New creates a new v1 ShellClient from an existing websocket.
+func New(ws *websocket.Conn) *ShellClient {
+	stdinReader, stdin := InfinitePipe(nil)
+	outputChan := make(chan bool)
+	stdout, stdoutWriter := DrainingPipe(MaxOutstandingBytes, outputChan)
+	stderr, stderrWriter := DrainingPipe(MaxOutstandingBytes, outputChan)
+
+	s := &ShellClient{
+		ws:      ws,
+		done:    make(chan struct{}),
+		success: true,
+
+		cPause: sync.NewCond(&sync.Mutex{}),
```

Comment:
Please create it as `m sync.Mutex` and `c sync.Cond` like you did in the pipes it's a lot prettier... and probably you can use just one mutex/cond and the entire object... that keeps it simpler even if they protect two variables (so long as it doesn't create a deadlock it's a good idea to have as few locks as possible).